src/epcpm/sunspectocsv.py

```diff
@@ -0,0 +1,532 @@
+    size=data_point_fields.size,
+    type=data_point_fields.type_uuid,
+    scale_factor=data_point_fields.factor_uuid,
+    units=data_point_fields.units,
+)
+
+bitfield_fields = Fields(
+    block_offset=bitfield_fields.block_offset,
+    size=bitfield_fields.size,
+    type=bitfield_fields.type_uuid,
+)
+
+
+def export(
+    path,
+    sunspec_model,
+    parameters_model,
+    column_filter=None,
+    skip_output=False,
+):
+    if skip_output:
+        return
+
+    if column_filter is None:
+        column_filter = attr_fill(Fields, True)
+
+    builder = epcpm.sunspectocsv.builders.wrap(
+        wrapped=sunspec_model.root,
+        parameter_uuid_finder=sunspec_model.node_from_uuid,
+        parameter_model=parameters_model,
+        column_filter=column_filter,
+    )
+
+    csv_data = builder.gen()
+
+    with open(path.with_suffix(".csv"), "w", newline="") as csv_file:
+        csv_writer = csv.writer(csv_file, quoting=csv.QUOTE_ALL)
+        for data_row in csv_data:
+            csv_writer.writerow(data_row)
+
+
+@builders(epcpm.sunspecmodel.Root)
+@attr.s
+class Root:
+    wrapped = attr.ib()
+    column_filter = attr.ib()
+    parameter_uuid_finder = attr.ib(default=None)
+    parameter_model = attr.ib(default=None)
+    sort_models = attr.ib(default=False)
+
+    def gen(self):
+        csv_data = []
+
+        if self.sort_models:
+            children = sorted(
+                self.wrapped.children,
+                key=lambda child: (
+                    0 if isinstance(child, epcpm.sunspecmodel.Model) else 1,
+                    getattr(child, "id", math.inf),
+                ),
+            )
+        else:
+            children = list(self.wrapped.children)
+
+        model_offset = 2  # account for starting 'SunS'
+        for model in children:
+            if isinstance(model, epcpm.sunspecmodel.Table):
+                # TODO: for now, implement it soon...
+                continue
+
+            model_offset += builders.wrap(
+                wrapped=model,
+                csv_data=csv_data,
+                column_filter=self.column_filter,
+                padding_type=self.parameter_model.list_selection_roots[
+                    "sunspec types"
+                ].child_by_name("pad"),
+                parameter_uuid_finder=self.parameter_uuid_finder,
+                model_offset=model_offset,
+            ).gen()
+
+        return csv_data
+
+
+@builders(epcpm.sunspecmodel.Model)
+@attr.s
+class Model:
+    wrapped = attr.ib()
+    csv_data = attr.ib()
+    column_filter = attr.ib()
+    padding_type = attr.ib()
+    model_offset = attr.ib()  # starting Modbus address for the model
+    parameter_uuid_finder = attr.ib(default=None)
+
+    def gen(self):
+        self.wrapped.children[0].check_offsets_and_length()
+
+        overall_length = sum(
+            child.check_offsets_and_length() for child in self.wrapped.children[1:]
+        )
+
+        add_padding = (overall_length % 2) == 1
```

Comment:
magic numbers / add comment

Reply:
Comment added in 004cc5dfe8e5765d8f063afd11dec72e1e9f2840